rct229/rule_engine/rct_report.py

```diff
@@ -0,0 +1,139 @@
+            for outcome in outcomes:
+                rule_outcome_dict = {
+                    RCTOutcomeLabel.PASS: 0,
+                    RCTOutcomeLabel.FAILED: 0,
+                    RCTOutcomeLabel.UNDETERMINED: 0,
+                    RCTOutcomeLabel.NOT_APPLICABLE: 0,
+                }
+                rule_report = self.generate_rule_report(outcome, rule_outcome_dict)
+                rule_outcome = self.calculate_rule_outcome(rule_outcome_dict)
+                self.add_rule_to_ruleset_report(
+                    ruleset_report, rule_report, rule_outcome
+                )
+            report_dir = os.path.join(report_dir, self.ruleset_report_file)
+            self.save_ruleset_report(ruleset_report, report_dir)
+
+    def generate_rule_report(self, rule_outcome, outcome_dict):
+        """
+        Function to generate a rule's report
+
+        Parameters
+        ----------
+        rule_outcome: json contains a rule's raw report
+        outcome_dict: outcome dictionary, default is
+
+        Returns: rule re
+        -------
+
+        """
+        return rule_outcome
+
+    def initialize_ruleset_report(self):
+        """
+        Initialize a data structure for generating a ruleset report
+        default is list
+
+        Returns: data structure, default is list
+        -------
+
+        """
+        return list()
+
+    def add_rule_to_ruleset_report(self, ruleset_report, rule_report, rule_outcome):
+        """
+        Add rule to the ruleset report
+
+        Parameters
+        ----------
+        ruleset_report: type depends on the initialized ruleset report
+        rule_report: type depends on the intialized ruleset re
+        rule_outcome: dict, contains counts for four RCT outputs
+        -------
+
+        """
+        pass
+
+    def calculate_rule_outcome(self, outcome_dict):
+        """
+        Calculate the outcome of a rule - the outcome should always match
+        to std 229 requirements.
+
+        Parameters
+        ----------
+        outcome_dict: dict outcomes dictionary
+
+        Returns: str outcome
+        -------
+
+        """
+        outcome_label = RCTOutcomeLabel.NOT_APPLICABLE
+        if outcome_dict[RCTOutcomeLabel.UNDETERMINED] > 0:
+            outcome_label = RCTOutcomeLabel.UNDETERMINED
+        elif outcome_dict[RCTOutcomeLabel.FAILED] > 0:
+            outcome_label = RCTOutcomeLabel.FAILED
+        elif outcome_dict[RCTOutcomeLabel.PASS] > 0:
+            outcome_label = RCTOutcomeLabel.PASS
+        return outcome_label
+
+    def save_ruleset_report(self, ruleset_report, report_dir):
+        """
+        Save the ruleset report to a file.
+
+        Parameters
+        ----------
+        ruleset_report: user-defined data structure (should be consistent with the data defined in initialize_ruleset_report function
+        report_dir: report file path.
+
+        Returns NA
+        -------
+
+        """
+        pass
```

Comment:
It may become clear as I continue my review, but at this point, it isn't clear why an implementation couldn't be given here.

Indeed, I now see that the report is not always a json file, so saving may need to be handled differently.

Reply:
Yes, that was the intent to have this function - The AHJ summary will be saved as strings so we will need this function to handle json and strings differently.